## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 553322`
Commit: `872faf980d653b608f84db57122a501a13286f69`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>python313Packages.nemo-relay</li>
    <li>python314Packages.nemo-relay</li>
  </ul>
</details>


---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>python313Packages.nemo-relay</li>
    <li>python314Packages.nemo-relay</li>
  </ul>
</details>


---
### `aarch64-darwin`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>python313Packages.nemo-relay</li>
    <li>python314Packages.nemo-relay</li>
  </ul>
</details>


---
<details>
<summary>python313Packages.nemo-relay</summary>
<pre>            self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        # Since Linux 6.12.9, SO_REUSEPORT is not allowed
        # on other address families than AF_INET/AF_INET6.
        if (
            self.allow_reuse_port and hasattr(socket, &quot;SO_REUSEPORT&quot;)
            and self.address_family in (socket.AF_INET, socket.AF_INET6)
        ):
            self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
&gt;       self.socket.bind(self.server_address)
E       PermissionError: [Errno 1] Operation not permitted

/nix/store/8gg2n5r87vsyy8d78n3s90a9r929c14v-python3-3.13.15/lib/python3.13/socketserver.py:478: PermissionError
=========================== short test summary info ============================
FAILED python/tests/test_observability_plugin.py::TestObservabilityConfigHelpers::test_atof_stream_sink_snapshots_header_env - PermissionError: [Errno 1] Operation not permitted
FAILED python/tests/test_observability_plugin.py::TestObservabilityConfigHelpers::test_atof_stream_sink_dotted_fields_deliver_during_plugin_clear[preserve-expected_data0] - PermissionError: [Errno 1] Operation not permitted
FAILED python/tests/test_observability_plugin.py::TestObservabilityConfigHelpers::test_atof_stream_sink_dotted_fields_deliver_during_plugin_clear[replace_dots-expected_data1] - PermissionError: [Errno 1] Operation not permitted
FAILED python/tests/test_types.py::TestOpenTelemetryTypes::test_subscriber_exports_scope_and_mark_events_end_to_end - PermissionError: [Errno 1] Operation not permitted
FAILED python/tests/test_types.py::TestOpenTelemetryTypes::test_gen_ai_subscriber_exports_standardized_agent_span - PermissionError: [Errno 1] Operation not permitted
FAILED python/tests/test_types.py::TestOpenInferenceTypes::test_subscriber_exports_scope_and_mark_events_end_to_end - PermissionError: [Errno 1] Operation not permitted
=========== 6 failed, 560 passed, 17 skipped, 3 deselected in 11.80s ===========</pre>
</details>
<details>
<summary>python314Packages.nemo-relay</summary>
<pre>            self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        # Since Linux 6.12.9, SO_REUSEPORT is not allowed
        # on other address families than AF_INET/AF_INET6.
        if (
            self.allow_reuse_port and hasattr(socket, &quot;SO_REUSEPORT&quot;)
            and self.address_family in (socket.AF_INET, socket.AF_INET6)
        ):
            self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
&gt;       self.socket.bind(self.server_address)
E       PermissionError: [Errno 1] Operation not permitted

/nix/store/kmq6qdcg0yq6r4awf02dgq003n42r6zk-python3-3.14.7/lib/python3.14/socketserver.py:478: PermissionError
=========================== short test summary info ============================
FAILED python/tests/test_observability_plugin.py::TestObservabilityConfigHelpers::test_atof_stream_sink_snapshots_header_env - PermissionError: [Errno 1] Operation not permitted
FAILED python/tests/test_observability_plugin.py::TestObservabilityConfigHelpers::test_atof_stream_sink_dotted_fields_deliver_during_plugin_clear[preserve-expected_data0] - PermissionError: [Errno 1] Operation not permitted
FAILED python/tests/test_observability_plugin.py::TestObservabilityConfigHelpers::test_atof_stream_sink_dotted_fields_deliver_during_plugin_clear[replace_dots-expected_data1] - PermissionError: [Errno 1] Operation not permitted
FAILED python/tests/test_types.py::TestOpenTelemetryTypes::test_subscriber_exports_scope_and_mark_events_end_to_end - PermissionError: [Errno 1] Operation not permitted
FAILED python/tests/test_types.py::TestOpenTelemetryTypes::test_gen_ai_subscriber_exports_standardized_agent_span - PermissionError: [Errno 1] Operation not permitted
FAILED python/tests/test_types.py::TestOpenInferenceTypes::test_subscriber_exports_scope_and_mark_events_end_to_end - PermissionError: [Errno 1] Operation not permitted
=========== 6 failed, 560 passed, 17 skipped, 3 deselected in 15.75s ===========</pre>
</details>

--- a/pkgs/development/python-modules/nemo-relay/default.nix
+++ b/pkgs/development/python-modules/nemo-relay/default.nix
@@ -1,0 +1,136 @@
+{
+  lib,
+  buildPythonPackage,
+  fetchFromGitHub,
+  rustPlatform,
+
+  # optional-dependencies
+  # langchain:
+  langchain,
+  langchain-core,
+  langgraph,
+  langgraph-checkpoint,
+  langsmith,
+  # deepagents:
+  langchain-anthropic,
+  # langchain-nvidia:
+  aiohttp,
+
+  # tests
+  cacert,
+  opentelemetry-proto,
+  pydantic,
+  pytest-asyncio,
+  pytestCheckHook,
+  typing-extensions,
+}:
+
+buildPythonPackage (finalAttrs: {
+  pname = "nemo-relay";
+  version = "0.7.3";
+  pyproject = true;
+  __structuredAttrs = true;
+
+  src = fetchFromGitHub {
+    owner = "NVIDIA";
+    repo = "nemo-relay";
+    tag = finalAttrs.version;
+    hash = "sha256-g7xHQOcccuyHIBiVY5GQHpd1vk99RMwuw923OR4+x3E=";
+  };
+
+  postPatch = ''
+    substituteInPlace Cargo.toml \
+      --replace-fail '"0.7.0"' '"${finalAttrs.version}"'
+  '';
+
+  cargoDeps = rustPlatform.fetchCargoVendor {
+    inherit (finalAttrs)
+      pname
+      version
+      src
+      ;
+    hash = "sha256-Re/R/0aSxFNNG9jnbSg+3D0OhQV1mPyxmIJT7ExFaP0=";
+  };
+
+  nativeBuildInputs = [
+    rustPlatform.cargoSetupHook
+    rustPlatform.maturinBuildHook
+  ];
+
+  optional-dependencies = {
+    cli = [
+      # no-op, placeholder for platform specific wheels
+    ];
+    langchain = [
+      langchain
+      langchain-core
+      langgraph
+      langgraph-checkpoint
+      langsmith
+    ];
+    langgraph = [
+      langgraph
+    ]
+    ++ finalAttrs.passthru.optional-dependencies.langchain;
+    deepagents = [
+      # deepagents # not packaged
+      langchain-anthropic
+    ]
+    ++ finalAttrs.passthru.optional-dependencies.langgraph;
+    langchain-nvidia = [
+      # langchain-nvidia-ai-endpoints # not packaged
+      aiohttp
+    ]
+    ++ finalAttrs.passthru.optional-dependencies.langchain;
+  };
+
+  preCheck = ''
+    rm -r python/nemo_relay
+  '';
+
+  nativeCheckInputs = [
+    cacert
+    opentelemetry-proto
+    pydantic
+    pytest-asyncio
+    pytestCheckHook
+    typing-extensions
+  ]
+  ++ finalAttrs.passthru.optional-dependencies.langchain;
+
+  disabledTestPaths = [
+    # wants to call uv
+    "python/tests/plugin/test_package_build.py::test_sdist_rebuilds_worker_bindings_without_checked_in_codegen"
+    # wants to call cargo
+    "python/tests/test_dynamic_plugin_host.py::test_worker_activation_executes_and_releases_callbacks"
+    "python/tests/test_dynamic_plugin_host.py::test_worker_activation_finalizer_never_waits_on_python_thread"
+  ];
+
+  pythonImportsCheck = [
+    "nemo_relay"
+    "nemo_relay._native"
+    # taken from https://github.com/NVIDIA/NeMo-Relay/tree/main/python/nemo_relay#package-surface
+    "nemo_relay.adaptive"
+    "nemo_relay.codecs"
+    "nemo_relay.guardrails"
+    "nemo_relay.intercepts"
+    "nemo_relay.llm"
+    "nemo_relay.observability"
+    "nemo_relay.plugin"
+    "nemo_relay.scope"
+    "nemo_relay.subscribers"
+    "nemo_relay.tools"
+    "nemo_relay.typed"
+  ];
+
+  __darwinAllowLocalNetworking = true;
+
+  meta = {
+    description = "Python bindings for the NeMo Relay agent runtime";
+    homepage = "https://github.com/NVIDIA/NeMo-Relay";
+    changelog = "https://github.com/NVIDIA/NeMo-Relay/releases/tag/${finalAttrs.src.tag}";
+    maintainers = with lib.maintainers; [ SuperSandro2000 ];
+    teams = [ lib.teams.cuda ];
+    license = lib.licenses.asl20;
+  };
+})

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -11830,6 +11830,8 @@ self: super: with self; {
 
   nebula3-python = callPackage ../development/python-modules/nebula3-python { };
 
+  nemo-relay = callPackage ../development/python-modules/nemo-relay { };
+
   nemosis = callPackage ../development/python-modules/nemosis { };
 
   nengo = callPackage ../development/python-modules/nengo { };

